dataset/sampling.py

```diff
@@ -0,0 +1,169 @@
+        Y : numpy.ndarray of rank 2 or None
+            Numpy array for training labels. Its dtype should be float32, float64, or int64.
+            None if there is no label array.
+        file_path_prefix : str
+            The path and name of the resulting tfrecord file to be generated, without '.tfrecords'
+        verbose : bool
+            If true, progress is reported.
+
+        Raises
+        ------
+        ValueError
+            If input type is not float (64 or 32) or int.
+
+        """
+
+        def _dtype_feature(ndarray):
+            """match appropriate tf.train.Feature class with dtype of ndarray. """
+            assert isinstance(ndarray, np.ndarray)
+            dtype_ = ndarray.dtype
+            if dtype_ == np.float64 or dtype_ == np.float32:
+                return lambda array: tf.train.Feature(float_list=tf.train.FloatList(value=array))
+            elif dtype_ == np.int64:
+                return lambda array: tf.train.Feature(int64_list=tf.train.Int64List(value=array))
+            else:
+                raise ValueError("The input should be numpy ndarray. \
+                                   Instaed got {}".format(ndarray.dtype))
+
+        assert isinstance(X, np.ndarray)
+        assert len(X.shape) == 2  # If X has a higher rank,
+        # it should be rshape before fed to this function.
+        assert isinstance(Y, np.ndarray) or Y is None
+
+        # load appropriate tf.train.Feature class depending on dtype
+        dtype_feature_x = _dtype_feature(X)
+        if Y is not None:
+            assert X.shape[0] == Y.shape[0]
+            assert len(Y.shape) == 2
+            dtype_feature_y = _dtype_feature(Y)
+
+            # Generate tfrecord writer
+        result_tf_file = file_path_prefix + '.tfrecords'
+        writer = tf.python_io.TFRecordWriter(result_tf_file)
+        if verbose:
+            print "Serializing {:d} examples into {}".format(X.shape[0], result_tf_file)
+
+        # iterate over each sample,
+        # and serialize it as ProtoBuf.
+        for idx in range(X.shape[0]):
+            x = X[idx]
+            if Y is not None:
+                y = Y[idx]
+
+            d_feature = {}
+            d_feature['X'] = dtype_feature_x(x)
+            if Y is not None:
+                d_feature['Y'] = dtype_feature_y(y)
+
+            features = tf.train.Features(feature=d_feature)
+            example = tf.train.Example(features=features)
+            serialized = example.SerializeToString()
+            writer.write(serialized)
+
+        if verbose:
+            print "Writing {} done!".format(result_tf_file)
+
+    def process(self, length=32, skip=1):
+        X, Y = self.slice_all()
+        # problem: image type matters if bit 8 or bit 32 image
+        # https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html
+        x_reshaped = np.divide(X.reshape(X.shape[0], X.shape[1] * X.shape[2] * X.shape[3]), 100.0)
```

Comment:
here, original shape of X is (num_vids, num_frames_in_vid, frame_h, frame_w) & shape of Y is (num_vids, num_frames_in_vid, 2, frame_h, frame_w) right?

Reply:
This is wrong. We have to divide it by 50 and subtract 1(to bring it to -1, +1 interval from 0, 100 interval), not the entire thing by 100.

Reply:
Yes shape is correct. 
Will change the division forgot L doesn't have negative range